TypeCobol/Compiler/Scopes/Scope.cs

```diff
@@ -0,0 +1,378 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TypeCobol.Compiler.Symbols;
+
+namespace TypeCobol.Compiler.Scopes
+{
+    /// <summary>
+    /// A Cobol Scope which contains Cobol Symbols declared within a Scope
+    /// </summary>
+    public class Scope<T> : IEnumerable<T> where T : Symbol
+    {
+        /// <summary>
+        /// A Symbol Entry in this scope
+        /// </summary>
+        public class Entry : IEnumerable<T>
+        {
+            /// <summary>
+            /// The symbol
+            /// </summary>
+            public  T Symbol
+            {
+                get;
+                protected set;
+            }
+
+            /// <summary>
+            /// The count of symbol
+            /// </summary>
+            public virtual int Count => 1;
+
+            /// <summary>
+            /// Indexer
+            /// </summary>
+            /// <param name="i"></param>
+            /// <returns></returns>
+            public virtual T this [int i]
+            {
+                get
+                {
+                    System.Diagnostics.Debug.Assert(i == 0);
+                    return i == 0 ? Symbol : null;
+                }
+            }
+
+            /// <summary>
+            /// Check this symbol exists in this entry
+            /// </summary>
+            /// <param name="sym">The symbol to check</param>
+            /// <returns>true if yes, false otherwise</returns>
+            public virtual bool Exists(T sym)
+            {
+                foreach (var t in this)
+                {
+                    if (t == sym)
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            /// <summary>
+            /// Enumerator on symbols
+            /// </summary>
+            /// <returns></returns>
+            public IEnumerator<T> GetEnumerator()
+            {
+                for (int i = 0; i < Count; i++)
+                {
+                    yield return this[i];
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// One Symbol constructor.
+        /// </summary>
+        /// <param name="symbol"></param>
+        public Entry(T symbol)
+            {
+                System.Diagnostics.Debug.Assert(symbol != null);
+                Symbol = symbol;
+            }
+        protected Entry()
+            {
+
+            }
+        }
+
+        /// <summary>
+        /// A Multi symbol entry class that represent several symbol.
+        /// </summary>
+        public class MultiSymbols : Entry
```

Comment:
We can safely merge Entry and MultiSymbols because Entry implements IEnumerable<T>, so basically every Entry is a collection. Just add a boolean property on the class to easily distinguish between single entries and multiple ones (i.e. `IsUnique` ?)

Reply:
Symbol is always the first symbol if any, this is just a distinction between a single entry and multiple entries, so it can works like that, because end users should only see Entry and not MultiSymbols.